scylla/src/transport/locator/tablets.rs

```diff
@@ -0,0 +1,630 @@
+        self.tablet_for_token(token).map(|tablet| {
+            tablet
+                .replicas
+                .per_dc
+                .get(dc)
+                .map(|x| x.as_slice())
+                .unwrap_or(&[])
+        })
+    }
+
+    fn add_tablet(&mut self, tablet: Tablet) {
+        // Smallest `left_idx` for which `tablet.first_token` is LESS OR EQUAL to `tablet_list[left_idx].last_token`.
+        // It implies that `tablet_list[left_idx]` overlaps with `tablet` iff `tablet.last_token`
+        // is GREATER OR EQUAL to `tablet_list[left_idx].first_token`.
+        let left_idx = self
+            .tablet_list
+            .partition_point(|t| t.last_token < tablet.first_token);
+        // Smallest `right_idx` for which `tablet.last_token` is LESS than `tablet_list[right_idx].first_token`.
+        // It means that `right_idx` is the index of first tablet that is "to the right" of `tablet` and doesn't overlap with it.
+        // From this it follows that if `tablet_list[left_idx]` turns out to not overlap with `tablet`, then `left_idx == right_idx`
+        // and we won't remove any tablets because `tablet` doesn't overlap with any existing tablets.
+        let right_idx = self
+            .tablet_list
+            .partition_point(|t| t.first_token <= tablet.last_token);
+        self.tablet_list.drain(left_idx..right_idx);
+        self.tablet_list.insert(left_idx, tablet);
+    }
+}
+
+impl TabletsInfo {
+    pub(crate) fn new() -> Self {
+        Self {
+            tablets: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn tablets_for_table<'a>(&'a self, table: &TableSpec) -> Option<&'a TableTablets> {
+        let tbl = self.tablets.get(table);
+
+        tbl
+    }
+
+    pub(crate) fn add_tablet(&mut self, table: TableSpec, tablet: Tablet) {
+        self.tablets.entry(table).or_default().add_tablet(tablet)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        collections::HashMap,
+        net::{IpAddr, Ipv4Addr, SocketAddr},
+        sync::Arc,
+    };
+
+    use scylla_cql::{
+        frame::response::result::{ColumnType, CqlValue},
+        types::serialize::{value::SerializeCql, CellWriter},
+    };
+    use uuid::Uuid;
+
+    use crate::{
+        routing::Token,
+        transport::{
+            locator::tablets::{
+                RawTablet, RawTabletReplicas, TabletParsingError, CUSTOM_PAYLOAD_TABLETS_V1_KEY,
+                RAW_TABLETS_CQL_TYPE,
+            },
+            topology::PeerEndpoint,
+            Node, NodeAddr,
+        },
+    };
+
+    use super::{TableTablets, Tablet, TabletReplicas};
+
+    #[test]
+    fn test_raw_tablet_deser_empty() {
+        let custom_payload = HashMap::new();
+        assert!(RawTablet::from_custom_payload(&custom_payload).is_none());
+    }
+
+    #[test]
+    fn test_raw_tablet_deser_trash() {
+        let mut custom_payload = HashMap::new();
+        custom_payload.insert(CUSTOM_PAYLOAD_TABLETS_V1_KEY.to_string(), vec![1, 2, 3]);
```

Comment:
This function is so small that I don't see the advantage of that.

Reply:
> Using iterators + `collect()` to create a `HashMap` instead of `insert`s seems much cleaner to me, as it avoids mutability.

Why not just:
```rust
let custom_payload =
            HashMap::from([(CUSTOM_PAYLOAD_TABLETS_V1_KEY.to_string(), vec![1, 2, 3])]);
```